release/nightly_tests/dataset/groupby_benchmark.py

```diff
@@ -0,0 +1,71 @@
+import ray
+import argparse
+from typing import Callable
+
+import pandas as pd
+
+from benchmark import Benchmark
+
+
+def parse_args() -> argparse.Namespace:
+    parser = argparse.ArgumentParser()
+    parser.add_argument("path", type=str)
+    parser.add_argument("--keys", required=True, nargs="+")
+
+    consume_group = parser.add_mutually_exclusive_group(required=True)
+    consume_group.add_argument("--aggregate", choices=["count", "mean", "std"])
+    consume_group.add_argument("--map-groups", action="store_true")
+
+    return parser.parse_args()
+
+
+def main(args):
+    benchmark = Benchmark("groupby")
+    consume_fn = get_consume_fn(args)
+
+    def benchmark_fn():
+        ds = ray.data.read_parquet(args.path).groupby(args.keys)
+        consume_fn(ds)
+
+    benchmark.run_fn(str(vars(args)), benchmark_fn)
+    benchmark.write_result()
+
+
+def get_consume_fn(
+    args: argparse.Namespace,
+) -> Callable[["ray.data.grouped_data.GroupedData"], None]:
+    if args.aggregate:
+
+        def consume_fn(ds):
+            if args.aggregate == "count":
+                aggregate_ds = ds.count()
+            elif args.aggregate == "mean":
+                aggregate_ds = ds.mean()
+            elif args.aggregate == "std":
+                aggregate_ds = ds.std()
+            else:
+                assert False, f"Invalid aggregate argument: {args.aggregate}"
+
+            aggregate_ds.materialize()
+
+    elif args.map_groups:
+
+        def consume_fn(ds):
+            def normalize_group(group: pd.DataFrame) -> pd.DataFrame:
+                return (group - group.mean()) / group.std()
+
+            # Users typically use the 'pandas' batch format with 'map_groups'.
+            ds.map_groups(normalize_group, batch_format="pandas")
```

Comment:
Can we do arrow instead (so that we have a baseline for what standard, recommended flow looks like)?

Reply:
OK.

We don't currently explicitly recommend PyArrow anywhere, right? Not sure this makes sense to do given how frequently people use the pandas batch format, but should we emit a warning if you use pandas, and document the recommendation?

Reply:
Is there any reason to not just test both? If we expose both we could just benchmark both? If we explicitly prefer one to the other though feel like we should do the warning + document the recommendation.

Reply:
> Is there any reason to not just test both?

At least with our current implementation, I don't think there's enough value in testing both. If there's a regression with one batch format, there will almost certainly be a regression with the other

Reply:
Actually @alexeykudinkin, to normalize a group with PyArrow, you'd have to separately normalize each column, which I think is probably slower than the vectorized pandas implementation.

```python
# Function to normalize all columns in the PyArrow table
def normalize_group(group):
    # List to hold normalized columns
    normalized_columns = []

    # Iterate over each column and normalize
    for col_name in group.column_names:
        column = group[col_name]
        
        # Compute mean and stddev for the entire column
        mean = pc.mean(column)
        stddev = pc.stddev(column)
        
        # Normalize the column: (column - mean) / stddev
        normalized_column = (column - mean) / stddev
        
        # Append the normalized column to the list
        normalized_columns.append(normalized_column)
    
    # Rebuild the PyArrow table from the normalized columns
    normalized_table = pa.table(normalized_columns, schema=table.schema)
    
    return normalized_table
```

Would I still test using the "pyarrow" batch format? I don't have a strong opinion either way

Reply:
All these Arrow ops are implemented as vectorized C++ kernels as well

Reply:
> All these Arrow ops are implemented as vectorized C++ kernels as well

Yeah, I understand that Arrow vectorizes the operations per-column.

pandas computes the mean for the entire table (`pd.DataFrame.mean()`). But on second thought, I guess they probably compute the results per-column under-the-hood.